DataManagementSystem/DB/FileCatalogDB.py

```diff
@@ -262,62 +262,126 @@ def hasAccess( self, opType, paths, credDict ):
+  def changePathMode( self, paths, credDict, recursive = False ):
+    """ Bulk method to change Mode for the given paths
 
-  def changePathMode( self, lfns, credDict, recursive = False ):
-    """ Change the mode of the given list of paths
+        :param dict paths: dictionary < lfn : mode >
+        :param dict credDict: dictionary of the caller credentials
+        :param boolean recursive: flag to apply the operation recursively
     """
-    res = self._checkPathPermissions( 'changePathMode', lfns, credDict )
+
+    res = self._checkPathPermissions( 'changePathMode', paths, credDict )
     if not res['OK']:
       return res
     failed = res['Value']['Failed']
+    successful = {}
+    if res['Value']['Successful']:
+      result = self.__changePathFunction( res['Value']['Successful'], credDict,
+                                          self.dtree.changeDirectoryMode,
+                                          self.fileManager.changeFileMode,
+                                          recursive = recursive )
+      failed.update( result['Value']['Failed'] )
+      successful = result['Value']['Successful']
+    return S_OK( { 'Successful':successful, 'Failed':failed } )
+
+  def __changePathFunction( self, paths, credDict,
+                            change_function_directory,
+                            change_function_file,
+                            recursive = False ):
+    """ A generic function to change Owner, Group or Mode for the given paths
+
+        :param dict paths: dictionary < lfn : parameter_value >
+        :param dict credDict: dictionary of the caller credentials
+        :param function change_function_directory: function to change directory parameters
+        :param function change_function_file: function to change file parameters
+        :param boolean recursive: flag to apply the operation recursively
+    """
+    dirList = []
+    result = self.isDirectory( paths, credDict )
+    if not result['OK']:
+      return result
+    for di in result['Value']['Successful']:
+      if result['Value']['Successful'][di]:
+        dirList.append( di )
+    fileList = []
+    if len( dirList ) < len( paths ):
+      result = self.isFile( paths, credDict )
+      if not result['OK']:
+        return result
+      for fi in result['Value']['Successful']:
+        if result['Value']['Successful'][fi]:
+          fileList.append( fi )
 
-    # if no successful, just return
-    if not res['Value']['Successful']:
-      return S_OK( {'Successful':{}, 'Failed':failed} )
+    successful = {}
+    failed = {}
 
-    res = self.fileManager.changePathMode(res['Value']['Successful'],credDict, recursive)    
-    if not res['OK']:
-      return res
-    failed.update( res['Value']['Failed'] )
-    successful = res['Value']['Successful']
+    dirArgs = {}
+    fileArgs = {}
+
+    for path in paths:
+      if ( not path in dirList ) and ( not path in fileList ):
+        failed[path] = 'Path not found'
```

Comment:
A standard error message like "No such file or directory", or even better, a DError object would be perfect here :-)